src/Compilers/Core/Portable/CodeGen/SwitchIntegralJumpTableEmitter.SwitchBucket.cs

```diff
@@ -60,28 +82,88 @@ internal ulong BucketSize
+            // Computed jump costs the same or more expensive than a regular conditional branch.
+            // So compbined cost would be 2+ conditional branches.
+            // Based on benchmarks it is actually closer to 3
+            //
+            // On the other hand, handling a degenerate bucket requires only 1 compare (range check) 
+            // so crumbling one bucket into two degenerate ones, when possible, is advantageous.
+            //
+            // One relatively common case that may result in degenerate switches is in async/iterator 
+            // state machines when control flow must use cascading dispatch into try blocks
+            internal int DegenerateBucketSplit
             {
                 get
                 {
                     if (_startLabelIndex == _endLabelIndex)
                     {
-                        // single element bucket needs exactly one compare
-                        return 1;
+                        // single element bucket is already degenerate
+                        return 0;
+                    }
+
+                    var allLabels = this._allLabels;
+                    var split = 0;
+                    var lastConst = this.StartConstant;
+                    var lastLabel = allLabels[_startLabelIndex].Value;
+
+                    for(int idx = _startLabelIndex + 1; idx <= _endLabelIndex; idx++)
+                    {
+                        var switchLabel = allLabels[idx];
+
+                        if (lastLabel != switchLabel.Value ||
+                            !IsContiguous(lastConst, switchLabel.Key))
+                        {
+                            if (split != 0)
+                            {
+                                // found another discontinuity, so cannot be split
+                                return -1; 
+                            }
+
+                            split = idx;
+                            lastLabel = switchLabel.Value;
+                        }
+
+                        lastConst = switchLabel.Key;
                     }
 
-                    // dense switch will perform two branches (range check and the actual computed jump)
-                    // computed jump is more expensive than a regular conditional branch
-                    // based on benchmarks the combined cost seems to be closer to 3
-                    //
-                    // this also allows in the "mostly sparse" scenario to avoid numerous 
-                    // little switches with only 2 labels in them.
-                    return 3;
+                    return split;
                 }
             }
 
+            private bool IsContiguous(ConstantValue lastConst, ConstantValue nextConst)
+            {
+                if (!lastConst.IsNumeric || !nextConst.IsNumeric)
+                {
+                    return false;
+                }
+
+                return nextConst.Int64Value - lastConst.Int64Value == 1;
```

Comment:
Yes, I will just make it call GetBucketSize. 

---
In reply to: [93716277](https://github.com/dotnet/roslyn/pull/16054#discussion_r93716277) [](ancestors = 93716277)